Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTP1Connections.swift

```diff
@@ -0,0 +1,438 @@
+                    stats.leased += 1
+                } else if connectionState.isAvailable {
+                    stats.idle += 1
+                }
+            }
+            return stats
+        }
+
+        var isEmpty: Bool {
+            self.connections.isEmpty
+        }
+
+        var canGrow: Bool {
+            if self.overflowIndex < self.connections.endIndex {
+                return self.overflowIndex < self.maximumConcurrentConnections
+            } else {
+                return self.connections.endIndex < self.maximumConcurrentConnections
+            }
+        }
+
+        var startingGeneralPurposeConnections: Int {
+            var connecting = 0
+            for connectionState in self.connections {
+                if connectionState.isConnecting, connectionState.isBackingOff {
+                    connecting += 1
+                }
+            }
+            return connecting
+        }
+
+        mutating func createNewConnection(on eventLoop: EventLoop) -> Connection.ID {
+            let connection = HTTP1ConnectionState(connectionID: self.generator.next(), eventLoop: eventLoop)
+            self.connections.append(connection)
+            self.overflowIndex = self.connections.index(after: self.overflowIndex)
+            return connection.connectionID
+        }
+
+        mutating func createNewOverflowConnection(on eventLoop: EventLoop) -> Connection.ID {
+            preconditionFailure()
+        }
+
+        mutating func newHTTP1ConnectionCreated(_ connection: Connection) -> Connection.ID {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connection.id }) else {
+                preconditionFailure("There is a new connection, that we didn't request!")
+            }
+
+            self.connections[index].started(connection)
+            return connection.id
+        }
+
+        mutating func backoffNextConnectionAttempt(_ connectionID: Connection.ID) -> EventLoop {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to create a new connection, that we know nothing about?")
+            }
+
+            self.connections[index].failedToStart()
+            return self.connections[index].eventLoop
+        }
+
+        mutating func replaceFailedOrClosedConnection(_ connectionID: Connection.ID) -> (Connection.ID, EventLoop) {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to create a new connection, that we know nothing about?")
+            }
+
+            let newConnection = HTTP1ConnectionState(
+                connectionID: self.generator.next(),
+                eventLoop: self.connections[index].eventLoop
+            )
+
+            self.connections[index] = newConnection
+            return (newConnection.connectionID, newConnection.eventLoop)
```

Comment:
Should we first check if the connection is actually failed or closed?

Reply:
If a connection fails in startup or is closed, I currently don't bother to set its state. Since the connections are in an ever changing array this would mean iterating over the connections twice, or exposing the array index. Neither of which do I like right now. In the `MockConnectionPool` (the structure I use for testing), I very much verify the state on every step.